content/blog/2022/2022-09-25-node-urlShortener.md

```diff
@@ -0,0 +1,494 @@
+
+This tutorial will walk you through how to develop a URL shortener application using the `MERN stack`. Developing `REST APIs` with `Node.js` and `MongoDB` and then integrating them into `React.js` frontend applications. This is a common coding interview question and an excellent practice project for new `Node.js` and `React` developers.
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/nodejs/url-shortener" %}}
+
+## Urlshortner Behind the Scene
+A URL shortener service allows users to enter long URLs and generate shorter, unique URLs in return. When the user lookup generated short URLs, they are redirected to the original long URL content.
+
+First, a URL shortener service selects a short domain name as a placeholder. Such as `goo.gl` `tinyurl.com` or `bit.ly`  Then a  unique shortcode or id is generated. For example, the URL https://reflectoring.io/schedule-cron-job-in-node/ can be shortened to https://lnkd.in/dFZXFq2T.
+
+These generated shortcodes are stored alongside the domain name.
+
+And then mapped to the long URL.
+
+When we visit the newly generated short URL. A database lookup is performed using the shortcode as a key, and if it is found, we are redirected to the long URL's content.
+
+URL shortener service simply uses a URL redirect. However, scaling, monitoring, URL filtering, spam prevention, URL verification, and other required features can result in more advanced logics.
+
+## Setting up `Node.Js` Application
+
+To begin, move into a new root directory where we want our application to live.
+
+Enter the following command into the terminal:
+```
+mkdir urlbackend
+```
+And this:
+```
+cd urlbackend
+```
+The folder `urlbackend` is created here. Then we moved our terminal directory to `urlbackend`, which is where our `Node.js` server lives.
+
+Run the command below in the terminal to initialize our `Node.js` application.
+```
+npm init -y
+```
+Then open your `Node.js` application in your preferred IDE. 
+
+Using the terminal run the following command. Ensure you are in the `urlbackend` directory:
+
+```
+npm install cors dotenv express mongoose shortid
+```
+Here, we're setting up the required dependencies for our backend, which include:
+- cors: Cross-origin resource sharing (CORS) allows AJAX requests to skip the Same-origin policy and access resources from remote hosts. Comes in handy while connecting to our react application.
+- dotenv: Loads environment variables from a .env file into process.env
+- express: `Node.Js` framework
+- mongoose: An object modeling tool aids in connecting to and communicating with `MongoDB`.
+- shortid: Generates non-sequential short unique ids
+
+Next, create an `index.js` and a  `.env` file.
+
+Our application should be structured like this:
+{{% image alt="We want to deploy our application into a staging and a production environment" src="images/posts/designing-a-aws-cdk-project/serverapp.png" %}}
```

Comment:
All images aren't showing because the path is wrong. Please have a look at the [preview](https://deploy-preview-503--reflectoring.netlify.app/node-url-shortener/) and fix.